File: ios/Dinearound-app/Dinearound-app/Services/APIClient.swift
import Foundation

enum APIClientError: LocalizedError {
    case notConfigured
    case invalidResponse
    case server(String)
    case unauthorized

    var errorDescription: String? {
        switch self {
        case .notConfigured: return "API URL not configured"
        case .invalidResponse: return "Invalid server response"
        case .server(let message): return message
        case .unauthorized: return "Unauthorized"
        }
    }
}

struct APIUser: Codable {
    let id: String
    let email: String
    let displayName: String?
}

struct APIVisit: Codable {
    let id: String
    let userId: String
    let restaurantId: String?
    let restaurantName: String
    let visitDate: String
    let rating: Int
    let notes: String
    let createdAt: String?
    let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case restaurantId = "restaurant_id"
        case restaurantName = "restaurant_name"
        case visitDate = "visit_date"
        case rating, notes
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

struct APIMedia: Codable {
    let id: String
    let restaurantId: String
    let visitId: String?
    let kind: String
    let url: String
    let contentType: String?

    enum CodingKeys: String, CodingKey {
        case id, kind, url
        case restaurantId = "restaurant_id"
        case visitId = "visit_id"
        case contentType = "content_type"
    }
}

private struct EmptyResponse: Decodable { let success: Bool }

private struct ErrorPayload: Decodable { let success: Bool; let error: String? }

final class APIClient {
    static let shared = APIClient()

    private static let tokenKey = "dinearound-auth-token"

    private init() {}

    var baseURL: URL? { AppConfig.apiBaseURL }

    var isEnabled: Bool { AppConfig.isAPIEnabled }

    var authToken: String? {
        get { UserDefaults.standard.string(forKey: Self.tokenKey) }
        set {
            if let newValue {
                UserDefaults.standard.set(newValue, forKey: Self.tokenKey)
            } else {
                UserDefaults.standard.removeObject(forKey: Self.tokenKey)
            }
        }
    }

    func register(email: String, password: String, displayName: String? = nil) async throws -> APIUser {
        struct Body: Encodable {
            let email: String
            let password: String
            let displayName: String?
        }
        struct Response: Decodable {
            let success: Bool
            let token: String
            let user: APIUser
            let error: String?
        }
        let result: Response = try await request(
            path: "/api/auth/register",
            method: "POST",
            body: Body(email: email, password: password, displayName: displayName),
            auth: false
        )
        authToken = result.token
        return result.user
    }

    func login(email: String, password: String) async throws -> APIUser {
        struct Body: Encodable { let email: String; let password: String }
        struct Response: Decodable {
            let success: Bool
            let token: String
            let user: APIUser
            let error: String?
        }
        let result: Response = try await request(
            path: "/api/auth/login",
            method: "POST",
            body: Body(email: email, password: password),
            auth: false
        )
        authToken = result.token
        return result.user
    }

    func logout() {
        authToken = nil
    }

    func fetchVisits() async throws -> [APIVisit] {
        struct Response: Decodable { let success: Bool; let data: [APIVisit] }
        let result: Response = try await request(path: "/api/visits")
        return result.data
    }

    func createVisit(
        id: String,
        restaurantId: String?,
        restaurantName: String,
        visitDate: String,
        rating: Int,
        notes: String
    ) async throws -> APIVisit {
        struct Body: Encodable {
            let id: String
            let restaurantId: String?
            let restaurantName: String
            let visitDate: String
            let rating: Int
            let notes: String
        }
        struct Response: Decodable { let success: Bool; let data: APIVisit }
        let result: Response = try await request(
            path: "/api/visits",
            method: "POST",
            body: Body(
                id: id,
                restaurantId: restaurantId,
                restaurantName: restaurantName,
                visitDate: visitDate,
                rating: rating,
                notes: notes
            )
        )
        return result.data
    }

    func updateVisit(
        id: String,
        restaurantId: String?,
        restaurantName: String,
        visitDate: String,
        rating: Int,
        notes: String
    ) async throws {
        struct Body: Encodable {
            let restaurantId: String?
            let restaurantName: String
            let visitDate: String
            let rating: Int
            let notes: String
        }
        let _: EmptyResponse = try await request(
            path: "/api/visits/\(id)",
            method: "PATCH",
            body: Body(
                restaurantId: restaurantId,
                restaurantName: restaurantName,
                visitDate: visitDate,
                rating: rating,
                notes: notes
            )
        )
    }

    func deleteVisit(id: String) async throws {
        let _: EmptyResponse = try await request(path: "/api/visits/\(id)", method: "DELETE")
    }

    func fetchWishlistIds() async throws -> Set<String> {
        struct Response: Decodable { let success: Bool; let ids: [String] }
        let result: Response = try await request(path: "/api/wishlist")
        return Set(result.ids)
    }

    func addWishlist(restaurantId: String) async throws {
        struct Body: Encodable { let restaurantId: String }
        let _: EmptyResponse = try await request(
            path: "/api/wishlist",
            method: "POST",
            body: Body(restaurantId: restaurantId)
        )
    }

    func removeWishlist(restaurantId: String) async throws {
        let _: EmptyResponse = try await request(path: "/api/wishlist/\(restaurantId)", method: "DELETE")
    }

    func saveMenuItems(
        restaurantId: String,
        items: [(category: String, name: String, price: String)]
    ) async throws {
        struct Item: Encodable { let category: String; let name: String; let price: String }
        struct Body: Encodable { let items: [Item] }
        let _: EmptyResponse = try await request(
            path: "/api/restaurants/\(restaurantId)/menu-items",
            method: "POST",
            body: Body(items: items.map { Item(category: $0.category, name: $0.name, price: $0.price) })
        )
    }

    func uploadMedia(
        data: Data,
        fileName: String,
        mimeType: String,
        restaurantId: String,
        visitId: String?,
        kind: String
    ) async throws -> APIMedia {
        var urlRequest = URLRequest(url: try apiURL(path: "/api/media/upload"))
        urlRequest.httpMethod = "POST"
        if let token = authToken {
            urlRequest.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let boundary = UUID().uuidString
        urlRequest.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        func appendField(_ name: String, _ value: String) {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n".data(using: .utf8)!)
            body.append("\(value)\r\n".data(using: .utf8)!)
        }
        appendField("restaurantId", restaurantId)
        if let visitId { appendField("visitId", visitId) }
        appendField("kind", kind)

        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(fileName)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: \(mimeType)\r\n\r\n".data(using: .utf8)!)
        body.append(data)
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        urlRequest.httpBody = body

        let (responseData, response) = try await URLSession.shared.data(for: urlRequest)
        guard let http = response as? HTTPURLResponse else { throw APIClientError.invalidResponse }

        struct UploadResponse: Decodable { let success: Bool; let data: APIMedia; let error: String? }
        let decoded = try JSONDecoder().decode(UploadResponse.self, from: responseData)
        if http.statusCode == 401 { throw APIClientError.unauthorized }
        if http.statusCode >= 400 {
            throw APIClientError.server(decoded.error ?? "Upload failed")
        }
        return decoded.data
    }

    private func apiURL(path: String) throws -> URL {
        guard let base = baseURL else { throw APIClientError.notConfigured }
        let normalized = path.hasPrefix("/") ? String(path.dropFirst()) : path
        return base.appendingPathComponent(normalized)
    }

    private func request<T: Decodable>(
        path: String,
        method: String = "GET",
        body: (any Encodable)? = nil,
        auth: Bool = true
    ) async throws -> T {
        var urlRequest = URLRequest(url: try apiURL(path: path))
        urlRequest.httpMethod = method
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if auth, let token = authToken {
            urlRequest.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        if let body {
            urlRequest.httpBody = try JSONEncoder().encode(AnyEncodable(body))
        }

        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let http = response as? HTTPURLResponse else { throw APIClientError.invalidResponse }

        if http.statusCode == 401 { throw APIClientError.unauthorized }
        if http.statusCode >= 400 {
            let err = (try? JSONDecoder().decode(ErrorPayload.self, from: data))?.error
            throw APIClientError.server(err ?? "Request failed (\(http.statusCode))")
        }

        return try JSONDecoder().decode(T.self, from: data)
    }
}

private struct AnyEncodable: Encodable {
    private let encode: (Encoder) throws -> Void
    init(_ wrapped: any Encodable) {
        encode = wrapped.encode
    }
    func encode(to encoder: Encoder) throws { try encode(encoder) }
}

File: web/lib/api-client.ts
/**
 * DineAround API client — shared Cloudflare Worker backend (D1 + R2).
 * Falls back gracefully when NEXT_PUBLIC_API_URL is unset (local-only mode).
 */

const TOKEN_KEY = 'dinearound-auth-token';

export function getApiBaseUrl(): string | null {
  const url = process.env.NEXT_PUBLIC_API_URL?.replace(/\/+$/, '');
  return url || null;
}

export function isApiEnabled(): boolean {
  return Boolean(getApiBaseUrl());
}

export function getAuthToken(): string | null {
  if (typeof window === 'undefined') return null;
  return localStorage.getItem(TOKEN_KEY);
}

export function setAuthToken(token: string | null): void {
  if (typeof window === 'undefined') return;
  if (token) localStorage.setItem(TOKEN_KEY, token);
  else localStorage.removeItem(TOKEN_KEY);
}

export interface ApiUser {
  id: string;
  email: string;
  displayName?: string;
}

export interface ApiVisit {
  id: string;
  user_id: string;
  restaurant_id: string | null;
  restaurant_name: string;
  visit_date: string;
  rating: number;
  notes: string;
  created_at: string;
  updated_at: string;
}

export interface ApiRestaurant {
  id: string;
  name: string;
  address: string;
  cuisine: string;
  rating: number;
  priceLevel: number;
  priceSymbols?: string;
  emoji: string;
  lat?: number | null;
  lng?: number | null;
  distanceKm?: number | null;
}

export interface ApiMedia {
  id: string;
  restaurant_id: string;
  visit_id?: string | null;
  kind: string;
  content_type?: string;
  url: string;
  created_at?: string;
}

class ApiError extends Error {
  constructor(
    message: string,
    public status: number
  ) {
    super(message);
    this.name = 'ApiError';
  }
}

async function apiFetch<T>(
  path: string,
  options: RequestInit & { auth?: boolean } = {}
): Promise<T> {
  const base = getApiBaseUrl();
  if (!base) throw new ApiError('API URL not configured', 0);

  const headers = new Headers(options.headers);
  if (!headers.has('Content-Type') && options.body && !(options.body instanceof FormData)) {
    headers.set('Content-Type', 'application/json');
  }

  if (options.auth !== false) {
    const token = getAuthToken();
    if (token) headers.set('Authorization', `Bearer ${token}`);
  }

  const res = await fetch(`${base}${path}`, { ...options, headers });
  const json = await res.json().catch(() => ({}));

  if (!res.ok) {
    throw new ApiError(json.error || res.statusText || 'Request failed', res.status);
  }
  return json as T;
}

export async function apiRegister(email: string, password: string, displayName?: string) {
  const result = await apiFetch<{ success: boolean; token: string; user: ApiUser }>(
    '/api/auth/register',
    {
      method: 'POST',
      auth: false,
      body: JSON.stringify({ email, password, displayName }),
    }
  );
  setAuthToken(result.token);
  return result;
}

export async function apiLogin(email: string, password: string) {
  const result = await apiFetch<{ success: boolean; token: string; user: ApiUser }>(
    '/api/auth/login',
    { method: 'POST', auth: false, body: JSON.stringify({ email, password }) }
  );
  setAuthToken(result.token);
  return result;
}

export async function apiLogout() {
  const base = getApiBaseUrl();
  if (base) {
    await fetch(`${base}/api/auth/logout`, { method: 'POST', credentials: 'include' }).catch(() => {});
  }
  setAuthToken(null);
}

export async function apiGetVisits() {
  const result = await apiFetch<{ success: boolean; data: ApiVisit[] }>('/api/visits');
  return result.data;
}

export async function apiCreateVisit(visit: {
  id?: string;
  restaurantId?: string;
  restaurantName: string;
  visitDate: string;
  rating: number;
  notes: string;
}) {
  const result = await apiFetch<{ success: boolean; data: ApiVisit }>('/api/visits', {
    method: 'POST',
    body: JSON.stringify({
      id: visit.id,
      restaurantId: visit.restaurantId,
      restaurantName: visit.restaurantName,
      visitDate: visit.visitDate,
      rating: visit.rating,
      notes: visit.notes,
    }),
  });
  return result.data;
}

export async function apiUpdateVisit(
  id: string,
  visit: Partial<{
    restaurantId: string;
    restaurantName: string;
    visitDate: string;
    rating: number;
    notes: string;
  }>
) {
  const result = await apiFetch<{ success: boolean; data: ApiVisit }>(`/api/visits/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(visit),
  });
  return result.data;
}

export async function apiDeleteVisit(id: string) {
  await apiFetch<{ success: boolean }>(`/api/visits/${id}`, { method: 'DELETE' });
}

export async function apiGetWishlistIds() {
  const result = await apiFetch<{ success: boolean; ids: string[] }>('/api/wishlist');
  return result.ids;
}

export async function apiAddWishlist(restaurantId: string) {
  await apiFetch('/api/wishlist', {
    method: 'POST',
    body: JSON.stringify({ restaurantId }),
  });
}

export async function apiRemoveWishlist(restaurantId: string) {
  await apiFetch(`/api/wishlist/${restaurantId}`, { method: 'DELETE' });
}

export async function apiGetNearby(params?: {
  lat?: number;
  lng?: number;
  q?: string;
  cuisine?: string;
  price?: number;
}) {
  const qs = new URLSearchParams();
  if (params?.lat != null) qs.set('lat', String(params.lat));
  if (params?.lng != null) qs.set('lng', String(params.lng));
  if (params?.q) qs.set('q', params.q);
  if (params?.cuisine) qs.set('cuisine', params.cuisine);
  if (params?.price) qs.set('price', String(params.price));
  const query = qs.toString() ? `?${qs}` : '';
  const result = await apiFetch<{ success: boolean; data: ApiRestaurant[] }>(
    `/api/restaurants/nearby${query}`,
    { auth: false }
  );
  return result.data;
}

export async function apiGetMenuItems(restaurantId: string) {
  const result = await apiFetch<{
    success: boolean;
    data: { id: string; category: string; name: string; price: string }[];
  }>(`/api/restaurants/${restaurantId}/menu-items`, { auth: false });
  return result.data;
}

export async function apiSaveMenuItems(
  restaurantId: string,
  items: { category: string; name: string; price: string }[]
) {
  const result = await apiFetch<{
    success: boolean;
    data: { id: string; category: string; name: string; price: string }[];
  }>(`/api/restaurants/${restaurantId}/menu-items`, {
    method: 'POST',
    body: JSON.stringify({ items }),
  });
  return result.data;
}

export async function apiUploadMedia(params: {
  file: Blob;
  fileName: string;
  restaurantId: string;
  visitId?: string;
  kind: 'food_photo' | 'menu_photo' | 'menu_pdf';
}) {
  const form = new FormData();
  form.append('file', params.file, params.fileName);
  form.append('restaurantId', params.restaurantId);
  if (params.visitId) form.append('visitId', params.visitId);
  form.append('kind', params.kind);

  const result = await apiFetch<{ success: boolean; data: ApiMedia }>('/api/media/upload', {
    method: 'POST',
    body: form,
  });
  return result.data;
}

export function apiMediaUrl(mediaId: string): string | null {
  const base = getApiBaseUrl();
  if (!base) return null;
  return `${base}/api/media/${mediaId}`;
}

export async function apiListMedia(restaurantId: string, visitId?: string, kind?: string) {
  const qs = new URLSearchParams();
  if (visitId) qs.set('visitId', visitId);
  if (kind) qs.set('kind', kind);
  const query = qs.toString() ? `?${qs}` : '';
  const result = await apiFetch<{ success: boolean; data: ApiMedia[] }>(
    `/api/restaurants/${restaurantId}/media${query}`
  );
  return result.data;
}

/** Map API visit row → local store shape */
export function apiVisitToLocal(v: ApiVisit) {
  return {
    id: v.id,
    restaurantId: v.restaurant_id || '',
    restaurantName: v.restaurant_name,
    visitDate: v.visit_date,
    rating: v.rating,
    notes: v.notes,
    createdAt: v.created_at,
  };
}
